python/services/dataservice/dmod/dataservice/dataset_manager_collection.py

```diff
@@ -16,6 +16,9 @@ class DatasetManagerCollection:
         default_factory=dict, init=False
     )
 
+    def __hash__(self) -> int:
+        return id(self)
```

Comment:
Maybe there are some implications of _dataclass_ I'm forgetting/missing, but don't we essentially get the same functionality automatically for a user-defined class (i.e., equality and hash equality only for the same object)?

And beyond whether we get that for free, is that really what we want here?  I.e., why wouldn't we want any object with an equal dictionary of managers to test/hash equally?  

Those probably are questions for use cases that, in practice, aren't going to matter much, so the PR doesn't need to get held up while you add full hashing and equality (i.e. a TODO will suffice, if even that much is needed).  But since you seemed to go out of your way to add this, I wanted to at least ask about it.

Reply:
Thanks for pointing this out, @robertbartel! I should have added a docstring here to clarify the intent. A `dataclass` will only create a `__hash__` if it is "safe" to do so. That is to say that the internal fields of the `dataclass` are immutable. 

From [`dataclasses`](https://docs.python.org/3/library/dataclasses.html#dataclasses.dataclass) documentation:

>__hash__() is used by built-in [hash()](https://docs.python.org/3/library/functions.html#hash), and when objects are added to hashed collections such as dictionaries and sets. Having a __hash__() implies that instances of the class are immutable. Mutability is a complicated property that depends on the programmer’s intent, the existence and behavior of __eq__(), and the values of the eq and frozen flags in the @dataclass decorator.
>
>By default, @dataclass will not implicitly add a [__hash__()](https://docs.python.org/3/reference/datamodel.html#object.__hash__) method unless it is safe to do so. Neither will it add or change an existing explicitly defined __hash__() method. Setting the class attribute __hash__ = None has a specific meaning to Python, as described in the __hash__() documentation.

`__hash__` will not be generated for this class since a `dict` is not hashable. However, for all intents and purposes, we use this around the application like a singleton. To enable passing the same instance around to each `fastapi` handler that needs it, instead of using a global variable, we can use `fastapi`s dependency injection system to inject an instance into a handler. To ensure that we are passing the same instance around and not creating new instances, we decorate the injectable function with `functools.lru_cache` to effectively return the same instance each time the injectable is executed. However, to make this work, `functools.lru_cache` requires that arguments are hashable. This means if any _other_ injectable function _requires_ our injectable `DatasetManagerCollection` and it also is decorated with `lru_cache`, then `DatasetManagerCollection` _needs_ to be hashable. So, the simplest way to do this was just return the `id` reference to the class instance.

To illustrate the problem in [code](https://github.com/NOAA-OWP/DMOD/blob/01aab671f41466a7757b6fe543e37e286374aeb4/python/services/dataservice/dmod/dataservice/_injectable.py#L15):

```python
@lru_cache
def dataset_manager_collection() -> DatasetManagerCollection:
    """
    Service singleton `DatasetManagerCollection`.
    All handlers or service level background tasks that need a `DatasetManagerCollection` _should_
    use this / depend on it via FastApi's dependency injection system.
    """
    return DatasetManagerCollection()


@lru_cache
def data_derive_util(
    dmc: Annotated[DatasetManagerCollection, Depends(dataset_manager_collection)]
) -> DataDeriveUtil:
    """
    Service singleton `DataDeriveUtil`.
    All handlers or service level background tasks that need a `DataDeriveUtil` _should_
    use this / depend on it via FastApi's dependency injection system.
    """
    return DataDeriveUtil(dataset_manager_collection=dmc)
```

`data_derive_util` depends on `dataset_manager_collection` as an argument _and_ is decorated with `lru_cache`, so the return value of `dataset_manager_collection` must be hashable.


FastApi mentions this in their [docs](https://fastapi.tiangolo.com/advanced/settings/?h=lru_cache#creating-the-settings-only-once-with-lru_cache) as a solution to passing around globals.

Reply:
> A `dataclass` will only create a `__hash__` if it is "safe" to do so.

Ah, gotcha.

> `__hash__` will not be generated for this class since a dict is not hashable.

Yeah, like I said before, we shouldn't hold this up, but I did some rudimentary testing, and I think it ends up being the case that two _DatasetManagerCollection_ objects could test as equal but have different hashes.  That is a slightly more practical, though still not-often-encounter violation of the intent for hashing.   So let's make sure we add a TODO to come back to this later.

Reply:
Ah, I see what you are saying. I wasn't thinking about `__eq__`. Are you suggesting that I overload `__eq__` like I have `__hash__`?

Reply:
With a pragmatic lens, we don't ever compare `DatasetManagerCollection`s. We could instead just raise not implemented. Thoughts?

Reply:
Let's defer the conversation, but put something in place that reminds us to come back it.  As you say, we don't compare those right now.  We might later.  So we should be careful and thorough in thinking how this should be, but time is better spent on other things right now.

Reply:
I am going to leave it as is for now an add a `TODO` with reference https://github.com/NOAA-OWP/DMOD/issues/598

Reply:
Added a `TODO` and reffed the related issue